numpyro/infer/util.py

```diff
@@ -416,49 +415,82 @@ def single_chain_init(key):
+    by running model conditioned on latent samples from `posterior_samples`.
 
     .. warning::
-        The interface for the `predictive` function is experimental, and
+        The interface for the `Predictive` class is experimental, and
         might change in the future.
 
-    :param jax.random.PRNGKey rng: seed to draw samples
     :param model: Python callable containing Pyro primitives.
     :param dict posterior_samples: dictionary of samples from the posterior.
-    :param args: model arguments.
+    :param callable guide: optional guide to get posterior samples of sites not present
+        in `posterior_samples`.
+    :param dict params: dictionary of values for param sites of model/guide.
+    :param int num_samples: number of samples
     :param list return_sites: sites to return; by default only sample sites not present
         in `posterior_samples` are returned.
-    :param int num_samples: number of samples
-    :param kwargs: model kwargs.
+    :param bool parallel: whether to predict in parallel using JAX vectorized map :func:`jax.vmap`.
+
     :return: dict of samples from the predictive distribution.
     """
-    def single_prediction(rng, samples):
-        model_trace = trace(seed(condition(model, samples), rng)).get_trace(*args, **kwargs)
+    def __init__(self, model, posterior_samples=None, guide=None, params=None, num_samples=None,
+                 return_sites=None, parallel=False):
+        if posterior_samples is None and num_samples is None:
+            raise ValueError("Either posterior_samples or num_samples must be specified.")
+
+        posterior_samples = {} if posterior_samples is None else posterior_samples
+
+        for name, sample in posterior_samples.items():
+            batch_size = sample.shape[0]
+            if (num_samples is not None) and (num_samples != batch_size):
+                warnings.warn("Sample's leading dimension size {} is different from the "
+                              "provided {} num_samples argument. Defaulting to {}."
+                              .format(batch_size, num_samples, batch_size), UserWarning)
+            num_samples = batch_size
+
+        if num_samples is None:
+            raise ValueError("No sample sites in posterior samples to infer `num_samples`.")
+
         if return_sites is not None:
-            sites = return_sites
+            assert isinstance(return_sites, (list, tuple, set))
+
+        self.model = model
+        self.posterior_samples = {} if posterior_samples is None else posterior_samples
+        self.num_samples = num_samples
+        self.guide = guide
+        self.params = params
+        self.return_sites = return_sites
+        self.parallel = parallel
+
+    def _single_prediction(self, val, model_args=(), model_kwargs={}):
+        rng_key, samples = val
+        model_trace = trace(seed(condition(self.model, samples), rng_key)).get_trace(
```

Comment:
@neerajprad I would like to make this work for improper prior so probably `substitute` is better used here. WDYT?

Reply:
Sure, could you clarify what will be the difference in behavior?

Reply:
[condition](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/handlers.py#L274) only affects sample sites but I would like to condition on num_samples values (obtained from MCMC) of a `param` site too, so `substitute` would be better used here.